<a href="https://colab.research.google.com/github/PosgradoMNA/actividades-de-aprendizaje-A01793654/blob/main/Actividad5(IBM_Mod_2_DA).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Ciencia y analítica de datos
## Actividad 5
## Cursos de IBM
### Data Analysis With Python
### Modulo 2
#### Profesor: Jobish Vallikavungal Devassia
#### Profesor Tutor: Roberto Antonio Guevara González
#### Bernardo Mijangos Flores
####Matricula A01793654

# Laboratorio 2

## Objetivos:

#### Manejar valores faltantes
#### Formato de datos correcto
#### Estandarizar y normalizar datos

Data wrangling es el proceso de convertir datos del formato inicial a un formato que puede ser mejor para el análisis.

In [1]:
import pandas as pd
import matplotlib.pylab as plt

Descargamos el dataframe autos.csv de la URL "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Data%20files/auto.csv"

posteriormente los cargamos a la carpeta de google colab para poder utilizarlo

In [2]:
filename = "auto.csv"

In [5]:
# lista que contiene el nombre de los encabezados.

headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]

In [6]:
# uso del método read_csv() de Pandas para cargar los datos
# Se establece el parámetro "names" igual a los "headers" de la lista previa.

df = pd.read_csv(filename, names = headers)

In [ ]:
# mostramos la informacion
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

se puede observar que existen varios signos de interrogación en el dataset, esos son valores faltantes pueden dificultar nuestro análisis posterior.

¿Cómo trabajar con los datos faltantes?

Pasos:
* Identificar datos faltantes
* Tratar con datos faltantes
* Formato de datos correcto


# Identificar los datos faltantes

# Convertir "?" a NaN

En el dataset del automóvil, los datos faltantes vienen con el signo de interrogación "?". Reemplazamos "?" con NaN (No es un número).

usamos la función .reemplazar(A, B, inplace = True) para reemplazar A por B.

In [9]:
import numpy as np

# reemplazamos "?" por NaN

df.replace("?", np.nan, inplace = True)
df.head(5)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               NaN  alfa-romero       gas        std          two   
1          3               NaN  alfa-romero       gas        std          two   
2          1               NaN  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]


Evaluación de datos faltantes

Se usa las siguientes funciones para identificar valores faltantes.

Existen dos métodos para detectar datos faltantes:

* .isnull()
* .notnull()

La salida es un valor booleano que indica si el valor que se pasa al argumento es de hecho falta de datos.

In [10]:
missing_data = df.isnull()
missing_data.head(5)

symboling  normalized-losses   make  fuel-type  aspiration  num-of-doors  \
0      False               True  False      False       False         False   
1      False               True  False      False       False         False   
2      False               True  False      False       False         False   
3      False              False  False      False       False         False   
4      False              False  False      False       False         False   

   body-style  drive-wheels  engine-location  wheel-base  ...  engine-size  \
0       False         False            False       False  ...        False   
1       False         False            False       False  ...        False   
2       False         False            False       False  ...        False   
3       False         False            False       False  ...        False   
4       False         False            False       False  ...        False   

   fuel-system   bore  stroke  compression-ratio  horsepower  peak-rpm  \
0        False  False   False              False       False     False   
1        False  False   False              False       False     False   
2        False  False   False              False       False     False   
3        False  False   False              False       False     False   
4        False  False   False              False       False     False   

   city-mpg  highway-mpg  price  
0     False        False  False  
1     False        False  False  
2     False        False  False  
3     False        False  False  
4     False        False  False  

[5 rows x 26 columns]

"True" significa que el valor es un valor perdido, mientras que "False" significa que el valor no es un valor perdido.

Usando un bucle for podemos calcular rápidamente la cantidad de valores que faltan en cada columna.  En el cuerpo del bucle for, el método ".value_counts()" cuenta el número de valores "True".

In [11]:
for column in missing_data.columns.values.tolist():
    print(column)
    print (missing_data[column].value_counts())
    print("")    

symboling
False    205
Name: symboling, dtype: int64

normalized-losses
False    164
True      41
Name: normalized-losses, dtype: int64

make
False    205
Name: make, dtype: int64

fuel-type
False    205
Name: fuel-type, dtype: int64

aspiration
False    205
Name: aspiration, dtype: int64

num-of-doors
False    203
True       2
Name: num-of-doors, dtype: int64

body-style
False    205
Name: body-style, dtype: int64

drive-wheels
False    205
Name: drive-wheels, dtype: int64

engine-location
False    205
Name: engine-location, dtype: int64

wheel-base
False    205
Name: wheel-base, dtype: int64

length
False    205
Name: length, dtype: int64

width
False    205
Name: width, dtype: int64

height
False    205
Name: height, dtype: int64

curb-weight
False    205
Name: curb-weight, dtype: int64

engine-type
False    205
Name: engine-type, dtype: int64

num-of-cylinders
False    205
Name: num-of-cylinders, dtype: int64

engine-size
False    205
Name: engine-size, dtype: int64

fuel-system
Fa

En el resumen se observa que cada columna tiene 205 filas de datos y siete de las columnas contienen datos faltantes:

* "normalized-losses": 41 datos perdidos
* "num-of-doors": 2 datos perdidos
* "bore": 4 datos perdidos
* "stroke" : 4 datos perdidos
* "horsepower": 2 datos perdidos
* "peak-rpm": 2 datos perdidos
* "price": 4 datos perdidos

¿Cómo lidiar con los datos faltantes?

* Drop data
 * Elimina la fila
 * Elimina la columna
* Replace data
 * Reemplaza por media
 * Reemplaza por la frecuencia
 * Reemplaza basado en otras funciones

Las columnas completas deben eliminarse solo si la mayoría de las entradas de la columna están vacías.

Reemplazo por media:

* "normalized-losses": 41 datos perdidos
* "stroke": 4 datos perdidos
* "bore": 4 datos perdidos
* "horsepower": 2 datos perdidos
* "peak-rpm": 2 datos perdidos

Reemplazar por frecuencia:

* "num-of-doors": 2 datos faltantes, reemplazarlos con "cuatro"
  * Motivo: el 84% de los autos sedan son cuatro puertas. Dado que cuatro puertas es lo más frecuente, es más probable que ocurra

Eliminar toda la fila

* "precio": 4 datos faltantes, simplemente se elimina toda la fila
 * Motivo: el precio es lo que queremos predecir. Cualquier entrada de datos sin datos de precios no se puede utilizar para la predicción; por lo tanto, cualquier fila ahora sin datos de precios no es útil

In [12]:
# Calculo del valor medio de la columna "normalized-losses"

avg_norm_loss = df["normalized-losses"].astype("float").mean(axis=0)
print("Average of normalized-losses:", avg_norm_loss)

Average of normalized-losses: 122.0


In [13]:
# Reemplazamos NaN con el valor medio de la columna "normalized-losses"

df["normalized-losses"].replace(np.nan, avg_norm_loss, inplace=True)

In [14]:
# Calculo del valor medio para la columna "bore"

avg_bore=df['bore'].astype('float').mean(axis=0)
print("Average of bore:", avg_bore)

Average of bore: 3.3297512437810943


In [15]:
# Reemplazo de NaN con el valor medio de la columna "bore"

df["bore"].replace(np.nan, avg_bore, inplace=True)

## Pregunta 1:

Según el ejemplo anterior, reemplace NaN en la columna "stroke" con el valor medio.

In [16]:
# Calculo de la media para la columna "stroke"

avg_stroke = df["stroke"].astype("float").mean(axis = 0)
print("Media de la columna stroke:", avg_stroke)

Media de la columna stroke: 3.255422885572139


In [18]:
# Calculo del valor medio de la columan "horsepower"

avg_horsepower = df['horsepower'].astype('float').mean(axis=0)
print("Media de la columna horsepower:", avg_horsepower)

Media de la columna horsepower: 104.25615763546799


In [19]:
# Reemplazo de NaN con el valor medio de la columna "horsepower"

df['horsepower'].replace(np.nan, avg_horsepower, inplace=True)

In [20]:
# Calculo del valor medio de la columna "peak-rpm"

avg_peakrpm=df['peak-rpm'].astype('float').mean(axis=0)
print("Media de la columna peak rpm:", avg_peakrpm)

Media de la columna peak rpm: 5125.369458128079


In [21]:
# Reemplazo de NaN con el valor medio de la columna "peak-rpm"

df['peak-rpm'].replace(np.nan, avg_peakrpm, inplace=True)

In [23]:
# Se puede ver qué valores están presentes en una columna en particular

df['num-of-doors'].value_counts()

four    114
two      89
Name: num-of-doors, dtype: int64

In [24]:
# Se puede ver que cuatro puertas son el tipo más común.
# También podemos usar el método ".idxmax()" para calcular automáticamente el tipo más común:

df['num-of-doors'].value_counts().idxmax()

'four'

In [25]:
# reemplaza los valores faltantes de la columna "num-of-doors" por la frecuencia 

df["num-of-doors"].replace(np.nan, "four", inplace=True)

In [26]:
# Elimina la fila con el valor NaN en la columna "price"
df.dropna(subset=["price"], axis=0, inplace=True)

# Reestablece el indice
df.reset_index(drop=True, inplace=True)

In [27]:
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3             122.0  alfa-romero       gas        std          two   
1          3             122.0  alfa-romero       gas        std          two   
2          1             122.0  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

# Formato correcto

El último paso en la limpieza de datos es verificar y asegurarse de que todos los datos estén en el formato correcto (int, float, text u otro).

Con Pandas se usa:

.dtype() para comprobar el tipo de datos

.astype() para cambiar el tipo de datos

In [28]:
# Se usa una lista con los tipos de datos para cada columna.

df.dtypes

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

Como podemos ver arriba, algunas columnas no son del tipo de datos correcto. Las variables numéricas deben tener el tipo 'flotante' o 'int', y las variables con cadenas, como las categorías, deben tener el tipo 'objeto'.

Por ejemplo, las variables 'bore' y 'stroke' son valores numéricos que describen los motores, por lo que deberíamos esperar que sean del tipo 'float' o 'int'; sin embargo, se muestran como tipo 'objeto'.

Tenemos que convertir los tipos de datos a un formato adecuado para cada columna usando el método "astype()".

In [29]:
# Convierte el tipo de dato al formato adecuado

df[["bore", "stroke"]] = df[["bore", "stroke"]].astype("float")
df[["normalized-losses"]] = df[["normalized-losses"]].astype("int")
df[["price"]] = df[["price"]].astype("float")
df[["peak-rpm"]] = df[["peak-rpm"]].astype("float")

In [30]:
# Se actualiza la lista de las columnas después de la conversión.

df.dtypes

symboling              int64
normalized-losses      int64
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                 float64
stroke               float64
compression-ratio    float64
horsepower            object
peak-rpm             float64
city-mpg               int64
highway-mpg            int64
price                float64
dtype: object

# Estandarización de datos

Los datos generalmente se recopilan de diferentes agencias en diferentes formatos. (La estandarización de datos también es un término para un tipo particular de normalización de datos donde restamos la media y dividimos por la desviación estándar).

La estandarización es el proceso de transformar los datos en un formato común, lo que permite al investigador hacer una comparación significativa.

Ejemplo

Transformar mpg a L/100km:

En nuestro dataset, las columnas de consumo de combustible "city-mpg" y "highway-mpg" están representadas por unidades de mpg (millas por galón). Supongamos que estamos desarrollando una aplicación en un país que acepta el consumo de combustible con el estándar L/100km.

Tendremos que aplicar la transformación de datos para transformar mpg en L/100km.

In [31]:
df.head()

symboling  normalized-losses         make fuel-type aspiration  \
0          3                122  alfa-romero       gas        std   
1          3                122  alfa-romero       gas        std   
2          1                122  alfa-romero       gas        std   
3          2                164         audi       gas        std   
4          2                164         audi       gas        std   

  num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0          two  convertible          rwd           front        88.6  ...   
1          two  convertible          rwd           front        88.6  ...   
2          two    hatchback          rwd           front        94.5  ...   
3         four        sedan          fwd           front        99.8  ...   
4         four        sedan          4wd           front        99.4  ...   

   engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0          130         mpfi  3.47    2.68               9.0        111   
1          130         mpfi  3.47    2.68               9.0        111   
2          152         mpfi  2.68    3.47               9.0        154   
3          109         mpfi  3.19    3.40              10.0        102   
4          136         mpfi  3.19    3.40               8.0        115   

   peak-rpm city-mpg  highway-mpg    price  
0    5000.0       21           27  13495.0  
1    5000.0       21           27  16500.0  
2    5000.0       19           26  16500.0  
3    5500.0       24           30  13950.0  
4    5500.0       18           22  17450.0  

[5 rows x 26 columns]

In [32]:
# Convertimos mpg a L/100km con una operacion matematica (235 dividido por mpg)
df['city-L/100km'] = 235/df["city-mpg"]

# Revisamos la transformacion de la data 
df.head()

symboling  normalized-losses         make fuel-type aspiration  \
0          3                122  alfa-romero       gas        std   
1          3                122  alfa-romero       gas        std   
2          1                122  alfa-romero       gas        std   
3          2                164         audi       gas        std   
4          2                164         audi       gas        std   

  num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0          two  convertible          rwd           front        88.6  ...   
1          two  convertible          rwd           front        88.6  ...   
2          two    hatchback          rwd           front        94.5  ...   
3         four        sedan          fwd           front        99.8  ...   
4         four        sedan          4wd           front        99.4  ...   

   fuel-system  bore  stroke  compression-ratio horsepower peak-rpm  city-mpg  \
0         mpfi  3.47    2.68                9.0        111   5000.0        21   
1         mpfi  3.47    2.68                9.0        111   5000.0        21   
2         mpfi  2.68    3.47                9.0        154   5000.0        19   
3         mpfi  3.19    3.40               10.0        102   5500.0        24   
4         mpfi  3.19    3.40                8.0        115   5500.0        18   

  highway-mpg    price  city-L/100km  
0          27  13495.0     11.190476  
1          27  16500.0     11.190476  
2          26  16500.0     12.368421  
3          30  13950.0      9.791667  
4          22  17450.0     13.055556  

[5 rows x 27 columns]

# Pregunta 2:
De acuerdo con el ejemplo anterior, transforme mpg a L/100km en la columna "highway-mpg" y cambie el nombre de la columna a "highway-L/100km".

In [33]:
# cambiamos mpg a L/100km con una operacion matematica (235 dividido por mpg)
df["highway-mpg"] = 235/df["highway-mpg"]

# cambiamos el nombre de la columna "highway-mpg" por "highway-L/100km"
df.rename(columns={'highway-mpg':'highway-L/100km'}, inplace=True)

# revisamos los cambios 
df.head()

symboling  normalized-losses         make fuel-type aspiration  \
0          3                122  alfa-romero       gas        std   
1          3                122  alfa-romero       gas        std   
2          1                122  alfa-romero       gas        std   
3          2                164         audi       gas        std   
4          2                164         audi       gas        std   

  num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0          two  convertible          rwd           front        88.6  ...   
1          two  convertible          rwd           front        88.6  ...   
2          two    hatchback          rwd           front        94.5  ...   
3         four        sedan          fwd           front        99.8  ...   
4         four        sedan          4wd           front        99.4  ...   

   fuel-system  bore  stroke  compression-ratio horsepower peak-rpm  city-mpg  \
0         mpfi  3.47    2.68                9.0        111   5000.0        21   
1         mpfi  3.47    2.68                9.0        111   5000.0        21   
2         mpfi  2.68    3.47                9.0        154   5000.0        19   
3         mpfi  3.19    3.40               10.0        102   5500.0        24   
4         mpfi  3.19    3.40                8.0        115   5500.0        18   

  highway-L/100km    price  city-L/100km  
0        8.703704  13495.0     11.190476  
1        8.703704  16500.0     11.190476  
2        9.038462  16500.0     12.368421  
3        7.833333  13950.0      9.791667  
4       10.681818  17450.0     13.055556  

[5 rows x 27 columns]

# Normalización de datos

La normalización es el proceso de transformar valores de varias variables en un rango similar. Las normalizaciones típicas incluyen escalar la variable para que el promedio de la variable sea 0, escalar la variable para que la varianza sea 1 o escalar la variable para que los valores de la variable oscilen entre 0 y 1.

Ejemplo

Para demostrar la normalización, digamos que queremos escalar las columnas "largo", "ancho" y "alto".

Objetivo: quisiera normalizar esas variables para que su valor oscile entre 0 y 1

Enfoque: reemplazar el valor original por (valor original)/(valor máximo)

In [34]:
# reemplaza (valor original) por (valor original)/(maximo valor)

df['length'] = df['length']/df['length'].max()
df['width'] = df['width']/df['width'].max()

# Pregunta 3:

De acuerdo con el ejemplo anterior, normalice la columna "height".

In [36]:
# reemplaza al igual que el codigo anterior
df['height'] = df['height']/df['height'].max() 

# mostramos las columnas escaladas
df[["length","width","height"]].head()

length     width    height
0  0.811148  0.890278  0.816054
1  0.811148  0.890278  0.816054
2  0.822681  0.909722  0.876254
3  0.848630  0.919444  0.908027
4  0.848630  0.922222  0.908027

# Agrupación

La agrupacion es un proceso de transformación de variables numéricas continuas en 'bins' categóricos discretos para el análisis agrupado.

Ejemplo:

En nuestro conjunto de datos, "horsepower" es una variable de valor real que va de 48 a 288 y tiene 59 valores únicos. ¿Qué pasa si solo nos preocupamos por la diferencia de precio entre autos de alta potencia, potencia media y poca potencia (3 tipos)? ¿Podemos reorganizarlos en tres "contenedores" para simplificar el análisis?

Usaremos el método de pandas 'cut' para segmentar la columna 'horsepower' en 3 contenedores.

In [37]:
# ejemplo de agrupacion con pandas

df["horsepower"]=df["horsepower"].astype(int, copy=True)